From 8b85960ef3c8d59f2999f777fd9c8e9942b0ff13 Mon Sep 17 00:00:00 2001
From: Corey Minyard <cminyard@mvista.com>
Date: Sat, 4 May 2019 14:09:16 +0200
Subject: [PATCH 270/345] MIPS: OCTEON: Don't translate underlying GPIO irq
 bits.

The GPIO interrupt dispatching requires that nothing "bind" to the
underlying CIU bits corresponding to the GPIO pins so that they are
free for use by the GPIO interrupt domain.

Make the default xlat() function return -EINVAL for GPIO pins.

Signed-off-by: Corey Minyard <cminyard@mvista.com>
[david.daney@cavium.com added CIU2 and CIU3 parts]
Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/cavium-octeon/octeon-irq.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/mips/cavium-octeon/octeon-irq.c b/arch/mips/cavium-octeon/octeon-irq.c
index 557e645abb1e..94e8993f5d5d 100644
--- a/arch/mips/cavium-octeon/octeon-irq.c
+++ b/arch/mips/cavium-octeon/octeon-irq.c
@@ -1241,6 +1241,10 @@ static int octeon_irq_ciu_xlat(struct irq_domain *d,
 	ciu = intspec[0];
 	bit = intspec[1];
 
+	/* Don't map GPIO pins through this interface. */
+	if (ciu == 0 && (bit >= 16 && bit < 32))
+		return -EINVAL;
+
 	if (ciu >= dd->num_sum || bit > 63)
 		return -EINVAL;
 
@@ -1961,6 +1965,10 @@ static int octeon_irq_ciu2_xlat(struct irq_domain *d,
 	ciu = intspec[0];
 	bit = intspec[1];
 
+	/* Don't map GPIO pins through this interface. */
+	if (ciu == 7)
+		return -EINVAL;
+
 	*out_hwirq = (ciu << 6) | bit;
 	*out_type = 0;
 
@@ -2420,6 +2428,7 @@ int octeon_irq_ciu3_xlat(struct irq_domain *d,
 
 	intsn_major = hwirq >> 12;
 	switch (intsn_major) {
+	case 0x03: /* GPIO handled separately. */
 	case 0x04: /* Software handled separately. */
 		return -EINVAL;
 	default:
-- 
2.25.1

